pyrk/th_system.py

```diff
@@ -104,3 +107,262 @@ def custom(self, t_b, t_env, res):
+                env = self.comp_from_name(interface)
+                QconvBC = self.convBoundary(component,
+                                            t_b=component.T[t_idx].magnitude,
+                                            t_env=env.T[t_idx].magnitude,
+                                            h=d["h"],
+                                            R=d["R"])
+                to_ret -= QconvBC/cap
+            if component.heatgen:
+                to_ret += self.heatgenFVM(component, power, omegas)/cap
+            for interface, d in component.cond.iteritems():
+                env = self.comp_from_name(interface)
+                Qcond = self.conductionFVM(component, env, t_idx)
+                to_ret -= Qcond/cap
+            for interface, d in component.conv.iteritems():
+                env = self.comp_from_name(interface)
+                if isinstance(env, THSuperComponent):
+                    Tr = env.compute_tr(component.T[t_idx].magnitude,
+                                        env.sub_comp[-2].T[t_idx].magnitude)
+                    Qconv = self.convection(t_b=component.T[t_idx].magnitude,
+                                            t_env=Tr,
+                                            h=d['h'],
+                                            A=d['area'])
+                    assert (Qconv*(component.T[t_idx].magnitude-Tr)) >= 0, '''
+                    convection from %s to %s, from low temperature %f to
+                    high temperature %f is wrong: %f''' % (
+                        component.name, env.name, component.T[t_idx].magnitude,
+                        Tr, Qconv.magnitude)
+                else:
+                    Qconv = self.convection(t_b=component.T[t_idx].magnitude,
+                                            t_env=env.T[t_idx].magnitude,
+                                            h=d['h'],
+                                            A=d['area'])
+                    assert (Qconv*(component.T[t_idx]-env.T[t_idx])).magnitude >= 0, \
+                        '''convection from %s to %s, from low temperature %f to
+                    high temperature %f is wrong: %f''' % (
+                        component.name, env.name, component.T[t_idx].magnitude,
+                        env.T[t_idx].magnitude, Qconv.magnitude)
+                to_ret -= Qconv/cap/component.vol.magnitude
+            for name, d in component.adv.iteritems():
+                Qadv = self.advection(t_out=component.T[t_idx].magnitude*2.0 -
+                                      d['t_in'].magnitude,
+                                      t_in=d['t_in'].magnitude,
+                                      m_flow=d['m_flow'],
+                                      cp=d['cp'])
+                to_ret -= Qadv/cap/component.vol.magnitude
+                assert Qadv >= 0, '''at step %d, %s is heating the system
+                by %f watts??? Tin is %f, tout is %f, tcomp is %f''' % (
```

Comment:
Also, to be more helpful, the ??? should probably be replaced with " and negative values are invalid" or something.